src/charm.py

```diff
@@ -69,29 +71,65 @@ def _on_config_changed(self, event) -> None:
         self.model.unit.status = MaintenanceStatus(msg)
         self.configure_workload_container(event)
 
+    def _get_ip_type(self, address) -> str:
+        """Validate a string type: ipv4, ipv6 or invalid as IP address."""
+        try:
+            ip = ipaddress.ip_address(address)
+
+            if isinstance(ip, ipaddress.IPv4Address):
+                return "ipv4"
+            elif isinstance(ip, ipaddress.IPv6Address):
+                return "ipv6"
+        except ValueError:
+            return
+
+    def readlines_reverse(self, qfile):
+        with qfile:
+            qfile.seek(0, os.SEEK_END)
+            position = qfile.tell()
+            line = StringIO("")
+            while position >= 0:
+                qfile.seek(position)
+                next_char = qfile.read(1)
+                if next_char == "\n":
+                    yield line.getvalue()[::-1]
+                    line = StringIO("")
+                else:
+                    line.write(next_char)
+                position -= 1
+            yield line.getvalue()[::-1]
+
     def _report_visits_by_ip_action(self, event: ActionEvent) -> None:
-        """Handle the report-visits-by-ip action."""
-        results = self._report_visits_by_ip(event)
+        """Handle the report-visits-by-ip action.
+
+        Args:
+            event: the Juju action event fired when the action executes.
+        """
+        results = self._report_visits_by_ip()
         event.set_results({"ips": str(results)})
 
-    def _report_visits_by_ip(self, _) -> list[(int, str)]:
+    def filter_lines(self, file) -> list[str]:
+        for line in self.readlines_reverse(file):
+            line = line.split()
+            if len(line) > 1 and datetime.strptime(
+                line[3].lstrip("[").rstrip("]"), "%d/%b/%Y:%H:%M:%S"
+            ) > (datetime.now() - timedelta(minutes=20)):
+                yield line
+            else:
+                break
+
+    def get_ip(self, log_list) -> list[str]:
```

Comment:
This can also be changed to just be a function that operates on a single line and then use it in list comprehension or some other iterator processing:

```Python
def get_ip(line: str) -> str:
    return line.split([0])

ips = map(get_ip, lines_within_timedelta)
```

The above still contains problems, e.g., what if the line is empty?